include/nektar_interface/solver_base/partsys_base.hpp

```diff
@@ -150,92 +95,174 @@ class PartSysBase {
-      if (this->sycl_target->comm_pair.rank_parent == 0) {
-        nprint("Ignoring (duplicate?) call to init_output().");
-      }
-    } else {
-      // Create H5Part instance
-      this->h5part =
-          std::make_shared<H5Part>(fname, this->particle_group, args...);
-      this->h5part_exists = true;
-    }
-  }
+  inline void init_output(std::string fname, T... args);
+
+  /// @brief Read some parameters associated with all particle systems.
+  inline void read_params();
 
   /**
    * @brief Store particle param values in a map.Values are reported later via
    * add_params_report()
-   *
    * @param label Label to attach to the parameter
    * @param value Value of the parameter that was set
    */
-  template <typename T> void report_param(std::string label, T val) {
-    // form stringstream and store string value in private map
-    std::stringstream ss;
-    ss << val;
-    param_vals_to_report[label] = ss.str();
-  }
+  template <typename T> inline void report_param(std::string label, T val);
 
+private:
+  /// Output frequency read from config file
+  int output_freq;
   /**
-   * @brief Read some parameters associated with all particle systems.
+   * Map containing parameter name,value pairs to be written to stdout when
+   * the nektar equation system is initialised. Populated with report_param().
    */
-  inline void read_params() {
-
-    // Read total number of particles / number per cell from config
-    int num_parts_per_cell, num_parts_tot;
-    this->session->LoadParameter(NUM_PARTS_TOT_STR, num_parts_tot, -1);
-    this->session->LoadParameter(NUM_PARTS_PER_CELL_STR, num_parts_per_cell,
-                                 -1);
-
-    if (num_parts_tot > 0) {
-      this->num_parts_tot = num_parts_tot;
-      if (num_parts_per_cell > 0) {
-        nprint("Ignoring value of '" + NUM_PARTS_PER_CELL_STR +
-               "' because  "
-               "'" +
-               NUM_PARTS_TOT_STR + "' was specified.");
-      }
-    } else {
-      if (num_parts_per_cell > 0) {
-        // Determine the global number of elements
-        const int num_elements_local = this->graph->GetNumElements();
-        int num_elements_global;
-        MPICHK(MPI_Allreduce(&num_elements_local, &num_elements_global, 1,
-                             MPI_INT, MPI_SUM, this->comm));
-
-        // compute the global number of particles
-        this->num_parts_tot =
-            ((int64_t)num_elements_global) * num_parts_per_cell;
-
-        report_param("Number of particles per cell/element",
-                     num_parts_per_cell);
-      } else {
-        nprint("Particles disabled (Neither '" + NUM_PARTS_TOT_STR +
-               "' or "
-               "'" +
-               NUM_PARTS_PER_CELL_STR + "' are set)");
-      }
-    }
-    report_param("Total number of particles", this->num_parts_tot);
+  std::map<std::string, std::string> param_vals_to_report;
+};
+
+inline PartSysBase::PartSysBase(const LU::SessionReaderSharedPtr session,
```

Comment:
If the definitions are split from the implementations is it worth putting them in a source cpp file in `src/nektar_interface/solver_base/partsys_base.cpp`?

Reply:
It would mean all the expensive to compile templated loops are done once.

Reply:
Good point, I was confusing NESOlib with NP and forgetting it wasn't header-only.  I'll split it up.

Reply:
Done